### ISSUE TITLE
flate2-rs unsound due to calls to `getenv` allowing safe data race

### ISSUE THREAD

**Issue by xerxes12354** [rustsec/advisory-db#926]
https://github.com/rust-lang/flate2-rs/issues/272

Closed without being addressed leaving security vulnerability. Unsound.

**Comment by thomcc** [comment]
This is honestly a bug in the stdlib that setenv is exposed in safe code.

Work is underway to potentially fix this in the stdlib, although it's tricky to find a way without breakage. In the meantime though, I'm kind of with Alex here that going "looking for FFI crates where this could be a problem" isn't exactly helpful 🤷

**Comment by tarcieri** [comment]
Based on a quick look I agree. We don't tend to track transitive unsoundness and would rather track a vuln against std for this.

If we don't already have one for this (I don't see one) I'd suggest filing a soundness vuln against std. But even that seems tricky as I don't see a canonical upstream tracking issue for this. This is the closest I've seen so far:

https://github.com/rust-lang/rust/issues/27970
https://github.com/rust-lang/rust/pull/24741/files

It seems like `set_var` is a huge footgun and perhaps unfixable in its current form, but there's not much consensus about what to do about it.

**Comment by xerxes12354** [comment]
You cannot create soundness bug using only Rust with this, `flate2-rs` allows this to occur by unsoundly exposing an unsafe API.

**Comment by thomcc** [comment]
Again, the bug here is in the stdlib exposing the setenv API as safe in the first place. I suspect the logic you're using is the one that was used at a time, but it's flawed: Rust doesn't exist in a vacuum. For years, C code has held that setenv is unsafe to use in the context of multiple threads, but that getenv is safe (this fact is even documented as part of POSIX). This API is too pervasively used to imagine that realistically Rust can assume safety given that the only users of it are from within Rust.

Note that in many cases, there's no viable solution here for C bindings. In some cases you can cut out the functionality or make it an `unsafe fn`, in others you can patch the original C library, but even that isn't always viable¹.

IMO, Rust should change the behavior of the `std::env::set*` functions to only impact the environment as viewed by `std::env::var*` (and by other rust-visible things like `std::process`), although possibly this should only take effect after the first thread is spawned.

Also, in practice, this isn't a problem 99% of the time, which is why despite the same issue being present in libstd, it still hasn't been fixed yet. Don't get me wrong: it should be fixed, but... realistically, nobody who isn't setting out to try and trigger this sort of bug will hit it. You have to aggressively perform set_var in a loop from a different thread than the reader, and that just isn't common behavior.

Again, this isn't to say that things are fine without a fix here, but just to say waiting on a fix from stdlib is probably find, rather than either arguing every C library binding.

---

¹ And even if patching is viable, e.g. in -sys crates that bundle/vendor it is honestly kinda dodgy. Personally, I would *really* rather random -sys crates I use not be applying various custom patches to the underlying C libs they're binding to, especially for something unlikely to cause issues like this.

**Comment by jhpratt** [comment]
> nobody who isn't setting out to try and trigger this sort of bug will hit it.

While I initially said the same thing, it turns out this isn't true. [Here's a user](https://github.com/time-rs/time/issues/296#issuecomment-734801026) saying that the bug (in time & chrono) was discovered in real-world usage.

Personally, I made the decision to eliminate the unsound behavior, though I am reconsidering that decision.

**Comment by pinkforest** [comment]
Sum summarum: flat2-rs unsoundness

There are upstream rust-lang/rust issue(s) and libs are well aware of this by now with extensive discussions.

There is also another issue more general in nature: https://github.com/rustsec/advisory-db/issues/1190

time was only re-filed because the maintainer wanted it back - https://github.com/rustsec/advisory-db/pull/1242#issuecomment-1144903688

chrono was only re-filed to advertise fix because there was RiR localtime_r in 0.4.20 - https://github.com/rustsec/advisory-db/pull/1310

Filing something against something where the flate2 maintainer disagrees that this should be done has made this controversial in addition that there can be no fix in flate2 itself.

Closing this issue in favor of the upstream issue https://github.com/rust-lang/rust/issues/90308 - if needed re: specific to flate2 - please re-open.

**Comment by FrankHB** [comment]
> Again, the bug here is in the stdlib exposing the setenv API as safe in the first place. I suspect the logic you're using is the one that was used at a time, but it's flawed: Rust doesn't exist in a vacuum. For years, C code has held that setenv is unsafe to use in the context of multiple threads, but that getenv is safe (this fact is even documented as part of POSIX). This API is too pervasively used to imagine that realistically Rust can assume safety given that the only users of it are from within Rust.

The logic is for various reasons in conventional all software development:

* This is the place to tracking *issues*, even not all of them are bugs.
* This is a public place searchable by all users, not necesarily of your project. Actually I come from Google result for some question s about `getenv` usability in multihtreading environment even not related to Rust. That does no makes the report in this issue useless.
* On the other hand, "Close" is a conventional primitive for all kinds of issue tracking. This includes the status of "CLOSED INVALID"/"CLOSED WONTFIX"/"CLOSED WORKSFORME"/"CLOSED DUP"/"CLOSED ERRATA"/... in traditional bugzilla instances. (Ironically, bugzilla is actually not just for "bugs"...) But instead of "the resolution is known and settled down", the reason you choose to close is just "we decide to have nothing to do further due to exteranl limitations", which is traditonally a separated top-level status like "SUSPENDED" or "ON HOLD".
* Sadly GitHub issues cannot reflect this need (and worse, the final actual is marked closed "as completed"). This is confusing in nature, because the remaining majority of developers will find the status is more closed to "OPEN", while you have no public documentation to override the commen sense of how to handle issues in general. If you choose to just inherit GitHub's issue model to ease of your life at the cost of inconvenience for arbitrary readers from public space, that is your freedom; but note that you will eventually reiterate your reasoning, as you are still the minority.

**Comment by RalfJung** [comment]
Please don't necro-post. The discussion here is long over, and if there's new things to discuss it should be done in a new issue.
